embassy-stm32/src/usb/otg.rs

```diff
@@ -498,72 +170,37 @@ impl<'d, T: Instance> embassy_usb_driver::Driver<'d> for Driver<'d, T> {
         max_packet_size: u16,
         interval_ms: u8,
     ) -> Result<Self::EndpointOut, EndpointAllocError> {
-        self.alloc_endpoint(ep_type, max_packet_size, interval_ms)
+        self.inner.alloc_endpoint(ep_type, max_packet_size, interval_ms)
     }
 
-    fn start(mut self, control_max_packet_size: u16) -> (Self::Bus, Self::ControlPipe) {
-        let ep_out = self
-            .alloc_endpoint(EndpointType::Control, control_max_packet_size, 0)
-            .unwrap();
-        let ep_in = self
-            .alloc_endpoint(EndpointType::Control, control_max_packet_size, 0)
-            .unwrap();
-        assert_eq!(ep_out.info.addr.index(), 0);
-        assert_eq!(ep_in.info.addr.index(), 0);
-
-        trace!("start");
+    fn start(self, control_max_packet_size: u16) -> (Self::Bus, Self::ControlPipe) {
+        let (bus, cp) = self.inner.start(control_max_packet_size);
 
         (
             Bus {
-                config: self.config,
                 phantom: PhantomData,
-                ep_in: self.ep_in,
-                ep_out: self.ep_out,
-                phy_type: self.phy_type,
+                inner: bus,
                 inited: false,
             },
-            ControlPipe {
-                _phantom: PhantomData,
-                max_packet_size: control_max_packet_size,
-                ep_out,
-                ep_in,
-            },
+            cp,
         )
     }
 }
 
 /// USB bus.
 pub struct Bus<'d, T: Instance> {
-    config: Config,
     phantom: PhantomData<&'d mut T>,
-    ep_in: [Option<EndpointData>; MAX_EP_COUNT],
-    ep_out: [Option<EndpointData>; MAX_EP_COUNT],
-    phy_type: PhyType,
+    inner: OtgBus<'d>,
     inited: bool,
 }
 
-impl<'d, T: Instance> Bus<'d, T> {
-    fn restore_irqs() {
-        T::regs().gintmsk().write(|w| {
-            w.set_usbrst(true);
-            w.set_enumdnem(true);
-            w.set_usbsuspm(true);
-            w.set_wuim(true);
-            w.set_iepint(true);
-            w.set_oepint(true);
-            w.set_rxflvlm(true);
-            w.set_srqim(true);
-            w.set_otgint(true);
-        });
-    }
-}
-
 impl<'d, T: Instance> Bus<'d, T> {
     fn init(&mut self) {
         super::common_init::<T>();
 
         // Enable ULPI clock if external PHY is used
-        let _ulpien = !self.phy_type.internal();
+        let phy_type = self.inner.phy_type();
+        let _ulpien = !phy_type.internal();
 
         #[cfg(any(stm32f2, stm32f4, stm32f7))]
         if T::HIGH_SPEED {
```

Comment:
maybe you can move all this stm32-specific stuff to `Driver::new_*`, so you don't need to make a wrapper for `Bus` at all

Reply:
I think I'd still need the wrapper because of the Drop implementation.

Reply:
ah, that's true :smiling_face_with_tear: 

I wonder if maybe generics is better for this in the end? I know yesterday I said i'd prefer no generics, but I was referring to "Instance" generics. Perhaps we can have a "flavor" trait instead, containing all the chip-specific stuff. embassy-stm32 and esp-hal would provide different impls. This would avoid the issues with the `Instance` generics: both stm32 instances would be `Otg<Stm32>`, not `Otg<OTG_FS>, Otg<OTG_HS>`. The generic param would stay inside the `Driver` wrapper, wouldn't get exposed to the user.

Reply:
We would end up with something like this in the driver, I believe:

```rust
/// Peripheral instance dependent USB IP configuration.
#[derive(Clone, Copy)]
pub struct OtgInstance<'d> {
    /// The USB peripheral.
    pub regs: Otg,
    /// The USB state.
    pub state: &'d State<{ MAX_EP_COUNT }>,
    /// The PHY type.
    pub phy_type: PhyType,
    /// Number of used endpoints.
    pub endpoint_count: usize,
}

pub trait Implementation {
    /// FIFO depth in words.
    const FIFO_DEPTH_WORDS: u16;

    /// Extra RX FIFO words needed by some implementations.
    const EXTRA_RX_FIFO_WORDS: u16;

    /// Whether to set up late cnak
    const QUIRK_SETUP_LATE_CNAK: bool;

    /// Calculates TRDT value based on some internal clock speed.
    fn calculate_trdt(instance: &OtgInstance, speed: vals::Dspd) -> u8;
}
```

I don't mind, I dislike lugging around hardware-specific constants in memory, but it's probably not as neat as you have in mind.

Reply:
yeah that's what I was hoping for :+1:

Reply:
Turns out, my predictive powers are terrible 😭 This all looked good on paper but the idea is not exactly feasible IMO.

`init` and `disable` both need the Instance to enable/disable interrupts at least, FIFO_DEPTH_WORDS is also Instance-specific.

What **could** end up in such a trait is `RX_FIFO_EXTRA_SIZE_WORDS` and `calculate_trdt` which doesn't really cry for a new trait.

Reply:
okay! no trait then :+1: